src/System.Buffers/src/System/Buffers/ArrayPoolBucket.cs

```diff
@@ -0,0 +1,104 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Diagnostics;
+using System.Threading;
+
+namespace System.Buffers
+{
+    /// <summary>
+    /// Provides a thread-safe bucket containing buffers that can be Rented and Returned as part 
+    /// of a buffer pool; it should not be used independent of the pool.
+    /// </summary>
+    internal sealed class ArrayPoolBucket<T>
+    {
+        private volatile int _index;
+        private T[][] _data;
+        private SpinLock _lock;
+
+        /// <summary>
+        /// Creates the pool with numberOfBuffers arrays where each buffer is of bufferLength length.
+        /// </summary>
+        internal ArrayPoolBucket(int bufferLength, int numberOfBuffers)
+        {
+            _index = 0;
+            _lock = new SpinLock();
+            _data = new T[numberOfBuffers][];
+            for (int i = 0; i < numberOfBuffers; i++)
+            {
+                _data[i] = new T[bufferLength];
+            }
+        }
+
+        /// <summary>
+        /// Returns an array from the Bucket sized according to the Bucket size.
+        /// If the Bucket is empty, null is returned.
+        /// </summary>
+        /// <returns>Returns a valid buffer when the bucket has free buffers; otherwise, returns null</returns>
+        internal T[] Rent()
+        {
+            T[] buffer;
+
+            // Use a SpinLock since it is super lightweight
+            // and our lock is very short lived
+            bool taken = false;
+            _lock.Enter(ref taken);
+            Debug.Assert(taken);
+
+            // Check if all of our buffers have been rented
+            if (_index >= _data.Length)
+            {
+                // Exit the lock as soon as possible
+                _lock.Exit(false);
+                return null;
+            }
+            else
```

Comment:
Rather than having two calls to Exit and two returns, I'd suggest changing the ```T[] buffer;``` earlier to be ```T[] buffer = null;``` and changing this if/else to instead be:
```C#
if (_index < _data.Length)
{
    buffer = _data[_index];
    _data[_index++] = null;
}
_lock.Exit(false);
return buffer;
```